### DISCUSSION TITLE
Increased color banding on export

### DISCUSSION THREAD

**Discussion by Ryudoadema** [nagadomi/nunif#157] [Q&A] [answered]
I am using a high quality source 1080p 8-bit h264. The export in fsbs is also 8-bit according to mediainfo. There is increased color banding on the export vs the original. Like in dark red to orange skies in dune or closeups of clothing the transitions are more linear and less smooth. Not by a huge margin, but noticeable. I've tried every setting I can think of to make it match up better to the source including 8 crf and slower preset (very little difference between the lower quality settings in the banding). Someone said iw3 exports in 16-bit, and with an 8-bit display you can't see the increased color depth. But mine all say 8-bit no matter the settings I use. And since the original 8-bit source looks smoother, something else must be going on. Any ideas?

**Comment by nagadomi** [comment, marked as answer]
About exported **images**, depth is 16-bit grayscale and rgb(frame image) is 8-bit RGB. 16-bit is related to depth and unrelated to rgb(frame image).
They are stored in PNG, a lossless format, so there are no compression artifacts or other degradation, but YUV->RGB colorspace conversion possibly changes the colors. (typical video is YUV420p  so YUV->RGB conversion is performed)

When converting from those images to SBS **video**, it is re-encoded with h264. the output video is 8bit (per channel, 24bpp).
Re-encoding will cause some degradation. It can be reduced by encoding settings, but it is impossible to eliminate it completely.
Also, if pixel format is yuv420p or yuv444p, RGB->YUV colorspace conversion is performed again.

There are two main reasons why the source video differs from the SBS video.
1. Re-encoding
2. Colorspace convertion

About re-encoding settings, I think CRF=8 and tune=grain is good enough.

About colorspace conversion, currently I am not doing anything special.
If someone can provide an example of a good ffmpeg option (video(YUV) -> image sequence(RGB) -> video(YUV)),
it may be able to support it.

**Reply by Ryudoadema** [reply]
Ah ok. So there is no way around the yuv to rgb to yuv colorspace conversions if selecting yuv420p with a yuv420p source. I think yuv420p is the most widely compatible, but I'll have to double-check on that. Thanks again.

**Reply by gituser123456789000** [reply]
> Ah ok. So there is no way around the yuv to rgb to yuv colorspace conversions if selecting yuv420p with a yuv420p source. I think yuv420p is the most widely compatible, but I'll have to double-check on that. Thanks again.

About compatibility, yeah, I'd stick with 420. I tried 444, but default video apps on my phone didn't play them.. or they only played sound. I had to download VLC player for the 444 videos to play, but once I noticed this and redid them in 420, they worked perfectly fine with the stock phone video players. I haven't tested the rbg option's compatibility.

**Reply by Ryudoadema** [reply]
> About exported **images**, depth is 16-bit grayscale and rgb(frame image) is 8-bit RGB. 16-bit is related to depth and unrelated to rgb(frame image). They are stored in PNG, a lossless format, so there are no compression artifacts or other degradation, but YUV->RGB colorspace conversion possibly changes the colors. (typical video is YUV420p so YUV->RGB conversion is performed)
> 
> When converting from those images to SBS **video**, it is re-encoded with h264. the output video is 8bit (per channel, 24bpp). Re-encoding will cause some degradation. It can be reduced by encoding settings, but it is impossible to eliminate it completely. Also, if pixel format is yuv420p or yuv444p, RGB->YUV colorspace conversion is performed again.
> 
> There are two main reasons why the source video differs from the SBS video.
> 
>     1. Re-encoding
> 
>     2. Colorspace convertion
> 
> 
> About re-encoding settings, I think CRF=8 and tune=grain is good enough.
> 
> About colorspace conversion, currently I am not doing anything special. If someone can provide an example of a good ffmpeg option (video(YUV) -> image sequence(RGB) -> video(YUV)), it may be able to support it.

So I just wanted to clarify that saying "export" in my question, I mean I am selecting "Full SBS" as my stereo format. I am not exporting disparity and re-importing in any way. The color change; banding; and granularity seen in that sample photo in your issue post #164 was from converting to "Full SBS" withing iw3 in one step. In that case, is the video still converting from yuv420; to rgb; and back again to yuv420? Just trying to see if I can find a fix. Thanks.

**Reply by nagadomi** [reply]
Yes, `Video -> RGB Image(Process Depth, SBS) -> Video` conversion is always performed.
I suspect that two causes,
1. limited color range(tv) -> full color range(pc) convertion
2. output YUV is not BT.709 and no metadata is added. 

However, I have not yet figured out how to handle these correctly in pyav.

If you check iw3's output mp4 file with mediainfo, you will see that the following information in the source video has disappeared.
```
Color range                              : Limited
Color primaries                          : BT.709
Transfer characteristics                 : BT.709
Matrix coefficients                      : BT.709
```

It's also difficult to test because it differs by video player.
https://www.mux.com/blog/your-browser-and-my-browser-see-different-colors

**Reply by Ryudoadema** [reply]
> Yes, `Video -> RGB Image(Process Depth, SBS) -> Video` conversion is always performed. I suspect that two causes,
> 
>     1. limited color range(tv) -> full color range(pc) convertion
> 
>     2. output YUV is not BT.709 and no metadata is added.
> 
> 
> However, I have not yet figured out how to handle these correctly in pyav.
> 
> If you check iw3's output mp4 file with mediainfo, you will see that the following information in the source video has disappeared.
> 
> ```
> Color range                              : Limited
> Color primaries                          : BT.709
> Transfer characteristics                 : BT.709
> Matrix coefficients                      : BT.709
> ```

Thanks. I am bypassing the issue completely at the moment by finishing the conversion in Resolve.

However, I have found some commands to bring the conversion in iw3 very close to the original source in color and reduce the banding issue a good amount. In the -vf box type put "eq=gamma=1.03, gradfun=1.6:24" without the quotes.

You can step the gamma up a little if you need more brightness to offset 3d dimming on tvs and projectors. I don't just to keep it closest to the source. Gradfun strength:radius. 1.6:16 or 1.6:24 is a good compromise of gradient smoothing with minimal detail loss. Strength can range from .51 to 64. Radius can be from 8 to 32. It really helps the issue a lot and if I were to use iw3 for the final conversion again this would definitely be part of it if the video include scenes that are prone to the issues.

**Reply by nagadomi** [reply]
Basically, if the color conversion is correct, the colors should not change.
If you feel you need more than that, the video player may be doing some post-processing.

**Reply by Ryudoadema** [reply]
Not sure then. Others are noticing the banding in their conversions as well so not just me at least.

I'm playing both files on the same players on the same pc simultaneously to check.  Only other possibility is if it's no longer picking up metadata that is now missing like you said. Selecting RGB as the pixel format is even worse.

**Reply by kaelsonofkrypto** [reply]
@3gMan @nagadomi i too am noticing banding, so nagadomi i can see why you wouldnt want to do 8 bit on depth maps, that makes sense, 16 bit is the way to go for sure. So why not make the rgb 16 bit also? wouldnt this further fix some of the issues to prevent banding? Thanks

**Reply by nagadomi** [reply]
@kaelsonofkrypto 
There is no advantage in most cases because most source images/videos are less than 8bit-RGB color (24bpp or MPEG Limited Range).
If the video is 10bit/12bit H.265, there is an advantage. However, it can only be displayed on a 10-bit PC monitor and proper Graphics Board settings.

----

Depth also makes no difference between 8-bit and 16-bit rendered image. But it is used as height in 3D space, so 16-bit depth is needed. See https://github.com/nagadomi/nunif/discussions/149#discussioncomment-9857930 for example.

**Reply by kaelsonofkrypto** [reply]
> @kaelsonofkrypto

Thanks. So if there is an advantage, for those who have a 10 bit monitor or projector and using cards like a rtx 3090 or 4090, because it is noticeable which i take it why some of us notice because some people doing 3d conversions will have high end equipment, Would you be able to give the option and this way you simply allow the user to choose default or the 16 bit for the rgb.  The goal would be to reduce degradation. Some people are using multiple programs, and Ideally u want the source to be the highest quality as possible when doing reconversions to achieve minimal degradation between conversions. For example some will use a manual deflicker in resolve because at the moment resolve does deflickering better. Some will do a motion interpolation in topaz to smooth out the motion. So this is why it would be important to have the higher quality option also. At least make it where the user can pick.

**Reply by nagadomi** [reply]
I may support it when all color/video format related issues are resolved, but I will not support it now.
Also, Python image library(PIL) does not support 16-bit RGB PNG, so it may be another image format such as EXR instead of PNG.
I would suggest you first check to see if the software you are using supports 16-bit RGB and what format it is.
